src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -569,6 +789,187 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
+/// For example, the pattern `-128...127i8` is encoded as `0..=255`.
+/// This makes comparisons and arithmetic on interval endpoints much more
+/// straightforward. See `signed_bias` for details.
+struct IntRange<'tcx> {
+    pub range: RangeInclusive<u128>,
+    pub ty: Ty<'tcx>,
+}
+
+impl<'tcx> IntRange<'tcx> {
+    fn from_ctor(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                 ctor: &Constructor<'tcx>)
+                 -> Option<IntRange<'tcx>> {
+        match ctor {
+            ConstantRange(lo, hi, end) => {
+                assert_eq!(lo.ty, hi.ty);
+                let ty = lo.ty;
+                let env_ty = ty::ParamEnv::empty().and(ty);
+                if let Some(lo) = lo.assert_bits(tcx, env_ty) {
+                    if let Some(hi) = hi.assert_bits(tcx, env_ty) {
+                        // Perform a shift if the underlying types are signed,
+                        // which makes the interval arithmetic simpler.
+                        let bias = IntRange::signed_bias(tcx, ty);
+                        let (lo, hi) = (lo ^ bias, hi ^ bias);
+                        // Make sure the interval is well-formed.
+                        return if lo > hi || lo == hi && *end == RangeEnd::Excluded {
+                            None
+                        } else {
+                            let offset = (*end == RangeEnd::Excluded) as u128;
+                            Some(IntRange { range: lo..=(hi - offset), ty })
+                        };
+                    }
+                }
+                None
+            }
+            ConstantValue(val) => {
+                let ty = val.ty;
+                if let Some(val) = val.assert_bits(tcx, ty::ParamEnv::empty().and(ty)) {
+                    let bias = IntRange::signed_bias(tcx, ty);
+                    let val = val ^ bias;
+                    Some(IntRange { range: val..=val, ty })
+                } else {
+                    None
+                }
+            }
+            Single | Variant(_) | Slice(_) => {
+                None
+            }
+        }
+    }
+
+    fn from_pat(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                pat: &Pattern<'tcx>)
+                -> Option<IntRange<'tcx>> {
+        Self::from_ctor(tcx, &match pat.kind {
+            box PatternKind::Constant { value } => ConstantValue(value),
+            box PatternKind::Range { lo, hi, end } => ConstantRange(lo, hi, end),
+            _ => return None,
+        })
+    }
+
+    // The return value of `signed_bias` should be XORed with an endpoint to encode/decode it.
+    fn signed_bias(tcx: TyCtxt<'_, 'tcx, 'tcx>, ty: Ty<'tcx>) -> u128 {
+        match ty.sty {
+            ty::TyInt(ity) => {
+                let bits = Integer::from_attr(tcx, SignedInt(ity)).size().bits() as u128;
+                1u128 << (bits - 1)
+            }
+            _ => 0
+        }
+    }
+
+    /// Convert a `RangeInclusive` to a `ConstantValue` or inclusive `ConstantRange`.
```

Comment:
nit: Convert an `IntRange`

Reply:
This function actually does take a `RangeInclusive` instead of an `IntRange`: it turned out to be more useful that way from its uses. Arguably it could not be on `IntRange` at all: but it fits in with the other logic, so I thought it made sense to be a function on `IntRange`.